RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc

```diff
@@ -0,0 +1,253 @@
+};
+
+PixelClusterTagInfoProducer::PixelClusterTagInfoProducer(const edm::ParameterSet& iConfig)
+    : m_jets(consumes<edm::View<reco::Jet> >(iConfig.getParameter<edm::InputTag>("jets"))),
+      m_vertices(consumes<reco::VertexCollection>(iConfig.getParameter<edm::InputTag>("vertices"))),
+      m_pixelhit(consumes<edmNew::DetSetVector<SiPixelCluster> >(iConfig.getParameter<edm::InputTag>("pixelhit"))),
+      m_isPhase1(iConfig.getParameter<bool>("isPhase1")),
+      m_addFPIX(iConfig.getParameter<bool>("addForward")),
+      m_minADC(iConfig.getParameter<int>("minAdcCount")),
+      m_minJetPt(iConfig.getParameter<double>("minJetPtCut")),
+      m_maxJetEta(iConfig.getParameter<double>("maxJetEtaCut")),
+      m_hadronMass(iConfig.getParameter<double>("hadronMass")),
+      m_nLayers(m_isPhase1 ? 4 : 3) {
+  produces<reco::PixelClusterTagInfoCollection>();
+}
+
+PixelClusterTagInfoProducer::~PixelClusterTagInfoProducer() {}
+
+// ------------ method called to produce the data  ------------
+void PixelClusterTagInfoProducer::produce(edm::StreamID iID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  // Declare produced collection
+  auto pixelTagInfo = std::make_unique<reco::PixelClusterTagInfoCollection>();
+
+  // Open jet collection
+  edm::Handle<edm::View<reco::Jet> > collectionJets;
+  iEvent.getByToken(m_jets, collectionJets);
+
+  // Count jets above threshold
+  int nJets(0);
+  for (auto jetIt = collectionJets->begin(); jetIt != collectionJets->end(); ++jetIt) {
+    if (jetIt->pt() > m_minJetPt)
+      nJets++;
+  }
+
+  // Get primary vertex in the event
+  edm::Handle<reco::VertexCollection> collectionPVs;
+  iEvent.getByToken(m_vertices, collectionPVs);
+  reco::VertexCollection::const_iterator firstPV = collectionPVs->begin();
+
+  // If no suitable Jet and PV is available, skip the event without opening pixel collection
+  if (collectionPVs->empty() || nJets <= 0) {
+    iEvent.put(std::move(pixelTagInfo));
+    return;
+  }
+
+  // Open Pixel Cluster collection
+  edm::Handle<edmNew::DetSetVector<SiPixelCluster> > collectionHandle;
+  iEvent.getByToken(m_pixelhit, collectionHandle);
+  const edmNew::DetSetVector<SiPixelCluster>& collectionClusters(*collectionHandle);
+
+  // Open Geometry
+  edm::ESHandle<TrackerGeometry> geom;
+  iSetup.get<TrackerDigiGeometryRecord>().get(geom);
+  const TrackerGeometry& theTracker(*geom);
+
+  // Retrieve tracker topology from geometry
+  edm::ESHandle<TrackerTopology> tTopoH;
+  iSetup.get<TrackerTopologyRcd>().get(tTopoH);
+  const TrackerTopology* tTopo = tTopoH.product();
+
+  std::vector<reco::PixelClusterProperties> clusters;
+
+  // Get vector of detunit ids, and fill a vector of PixelClusterProperties in the loop
+  for (auto const& detUnit : collectionClusters) {
+    if (detUnit.empty())
+      continue;
+    DetId detId = DetId(detUnit.detId());  // Get the Detid object
+    unsigned int detType = detId.det();    // det type, pixel = 1
+    if (detType != 1)
```

Comment:
"1" here appears like a magic number. Please replace L166 and L167 with the single
```C++
    if (detId.det() != DetId::Tracker)
```